### ISSUE TITLE
Add gradation curves

### ISSUE THREAD

**Issue by masc4ii** [ilia3101/MLV-App#136]
One more elements I am missing in MLVApp is gradation curve. Had already some fun with QtCreator... (functional prototype of GUI element, not more)
![bildschirmfoto 2018-11-24 um 10 38 48](https://user-images.githubusercontent.com/30245296/48966624-8ac46780-efd5-11e8-9a87-a236cc587862.png)
The processing shouldn't be too hard: calculate another 4 1D uint16 matrices and apply it. The parameter saving will be very ugly, because each point will have 2 parameters (limiting to 10 points per line we get 80 (4 * 2 * 10) more parameters... 😛 ).
No idea if my graphics element works on retina screens. 

**Edit**: the spline lib I used is not under GPL 😢 
https://www.atnf.csiro.au/computing/software/gipsy/sub/spline.c

**Comment by masc4ii** [comment]
Proof of concept
![bildschirmfoto 2018-11-24 um 16 33 48](https://user-images.githubusercontent.com/30245296/48970008-05f24180-f007-11e8-912d-b46878adcd1c.png)
...but still some flickering when values are changed while rendering...

**Comment by masc4ii** [comment]
Now with all 4 spline curves.
![bildschirmfoto 2018-11-24 um 17 10 22](https://user-images.githubusercontent.com/30245296/48970412-1822ae80-f00c-11e8-9672-149f1d457dcf.png)

**Comment by masc4ii** [comment]
Sure... I only don't like to upload code without license. Your lib looks good. All I need is the following:
`int spline1_c( float *xi, float *yi, int *nin, float *xo, float *yo, int *nout );`

Input: 
xi & yi - array of input points
nin - number of input points
xo - array of output points x coordinate
nout - number of output points

Output:
yo - array of output points y coordinate

Which of the functions could be the right one?!

**Comment by bouncyball-git** [comment]
There is also 'spline_prb.c' example. Couldn't it help to understand parameters?

**Comment by masc4ii** [comment]
Great... got it working... will cleanup a bit, then I'll commit! :-) Thanks so much!

**Comment by masc4ii** [comment]
https://github.com/ilia3101/MLV-App/commit/98721a74e06b4073d5bcb300821c3feef61d322e
Voilà. Have fun!

**Comment by bouncyball-git** [comment]
Cooool!

**Comment by masc4ii** [comment]
But something seems to be wrong with the new library:
![bildschirmfoto 2018-11-25 um 14 18 11](https://user-images.githubusercontent.com/30245296/48979558-241a7900-f0bd-11e8-8401-2a878c93fc2f.png)

**Comment by bouncyball-git** [comment]
Hmm... right, something's wrong there.

**Comment by masc4ii** [comment]
I tried out 2 other function sets of this library - seems also not to be correct. Each time different but not as expected.

**Comment by ilia3101** [comment]
Whoaa

**Comment by bouncyball-git** [comment]
Hehe yeah, lots of functions there. Maybe you need to fully understand the parameter meanings?

**Comment by masc4ii** [comment]
In the end the parameters are not so complicated:
`float *spline_cubic_set ( int n, float t[], float y[], int ibcbeg,
  float ybcbeg, int ibcend, float ybcend )`

    Input, int N, the number of data points.  N must be at least 2.
    In the special case where N = 2 and IBCBEG = IBCEND = 0, the
    spline will actually be linear.

    Input, float T[N], the knot values, that is, the points were data is
    specified.  The knot values should be distinct, and increasing.

    Input, float Y[N], the data values to be interpolated.

    Input, int IBCBEG, left boundary condition flag:
    0: the cubic spline should be a quadratic over the first interval;
    1: the first derivative at the left endpoint should be YBCBEG;
    2: the second derivative at the left endpoint should be YBCBEG;
    3: Not-a-knot: the third derivative is continuous at T(2).

    Input, float YBCBEG, the values to be used in the boundary
    conditions if IBCBEG is equal to 1 or 2.

    Input, int IBCEND, right boundary condition flag:
    0: the cubic spline should be a quadratic over the last interval;
    1: the first derivative at the right endpoint should be YBCEND;
    2: the second derivative at the right endpoint should be YBCEND;
    3: Not-a-knot: the third derivative is continuous at T(N-1).

    Input, float YBCEND, the values to be used in the boundary
    conditions if IBCEND is equal to 1 or 2.

    Output, float SPLINE_CUBIC_SET[N], the second derivatives
    of the cubic spline.

But whatever I try, something different happens but it looks not correct, while my other lib shows it perfectly without any parameter.

**Comment by bouncyball-git** [comment]
It's sad to hear :( maybe we ask GIPSY maintainers for it?

**Comment by bouncyball-git** [comment]
Yes I saw it  :)

**Comment by bouncyball-git** [comment]
I have one proposition if it's not so difficult to implement.

1. Double clicking or right clicking on spline point deletes it.
2. Double clicking on Y, R, G and B resets selected curve only.

Is it doable?

**Comment by bouncyball-git** [comment]
One more thing this tool can not recover data beyond what histogram shows right?

**Comment by masc4ii** [comment]
> Double clicking or right clicking on spline point deletes it.

Yes, I also thought this. Right now you can drag the points out of the diagram to delete them.

 > Double clicking or right clicking on Y, R, G and B resets selected curve only. 

I thought about another delete button. But how to separate between the two with icons? 

> this tool can not recover data beyond what histogram shows right?

Yapp, absolutely correct. That is sad, I know. The tool comes nealry in the very end of the processing part in the pipeline. So, all what was processed and limited before is already lost. But if we would add this tool at the beginning of the pipeline the behavior would be very strange.

**Comment by bouncyball-git** [comment]
> I thought about another delete button. But how to separate between the two with icons?

Button in QT does not have action like right mouse click right? If yes it is sad.

Interesting where curve processing is done in other applications.
Anyway, that is why I said to @ilia3101 recently that if we do not use -0-1+ floating point processing we have to go for 32bit or even 64bit integer internal processing path to have more room for processed values.

**Comment by bouncyball-git** [comment]
We really have to have second real double precision processing engine to compare how slower that would be. Maybe use some optimized for speed math libs? matrix operations etc...

**Comment by masc4ii** [comment]
I would have to write a special button for this... but I implemented it a bit different yesterday. Try it out. We also could delete both delete buttons and add a rightclick menu to the diagram.

I don't think that floating point processing will be any better here. The problem is still that the current architecture works with matrices for speed reasons. This must be changed when we want to have floating point. But when playing around I already saw with some parameters (Contrast, Clarity, Highlights, Shadows) that the math won't help when going over some limits... (don't know how to describe it better). This was nothing about our 0..65535 range; it was the end of the defined range of the mathematical function, which brought strange artifacts when going over the limits.
And when you try to calculate a histogram with on the 16bit data, you'll see that we don't have a resolution problem ;) (most colors are there on 0, 1 or 2 pixels only in the picture).
